modules/storages/app/contracts/storages/file_links/create_contract.rb

```diff
@@ -0,0 +1,52 @@
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) 2012-2022 the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+class Storages::FileLinks::CreateContract < BaseContract
+  validate :validate_storage_url
+  validate :validate_user_allowed_to_manage
+  validate :validate_project_storage_link
+
+  private
```

Comment:
There is no validation in here governing the "originData"-properties. Because of this, a client could create a file link with all of the properties being nil:

<img width="607" alt="image" src="https://user-images.githubusercontent.com/617519/158375621-dc5f3bc2-5444-49da-b54e-0102757ffc11.png">

I'd expect that at least the properties `id` and `name` can never be nil. The screenshot above was generated by not providing any `originData` at all. If `originData` is provided, the parser ensures that at least `createdAt` and `lastModifiedAt` are provided. But a request like:

<img width="655" alt="image" src="https://user-images.githubusercontent.com/617519/158376315-4cd0ac0c-3f5c-4cc0-88fb-19f868aece69.png">

will still lead to the file link being created:

<img width="589" alt="image" src="https://user-images.githubusercontent.com/617519/158376386-c8f16841-f268-42ec-9fce-772032598938.png">

It currently is also possible to create two file links to the same work package, having the same `originData/id` and the same `storageUrl` which I would consider a duplicate.

Reply:
And POST behavior implemented in dbbbd5b759 and f07f79389b.